src/snowflake/connector/network_async.py

```diff
@@ -0,0 +1,637 @@
+        if not self._is_active():
+            raise Exception(
+                "Runner should be active before getting loop, call start method first"
+            )
+
+        return self._loop
+
+    def start(self) -> None:
+        if self._is_active():
+            raise Exception("Runner is already active")
+
+        self._loop = asyncio.new_event_loop()
+        self._thread = Thread(
+            target=lambda loop: loop.run_forever(), args=(self._loop,)
+        )
+        self._thread.start()
+
+        # YICHUAN: Wait for event loop to actually start so no coroutines are run while thread is still working, this
+        # is a bit hacky but there's not much else we can do about it
+        while not self._is_active():
+            time.sleep(0.1)
+
+    def stop(self) -> None:
+        if not self._is_active():
+            raise Exception("Runner should be active before closing")
+
+        # YICHUAN: IT IS NOT SAFE TO ATTEMPT TO CLOSE A LOOP FROM ANOTHER (main) THREAD
+        self._loop.call_soon_threadsafe(self._loop.stop)
+        self._thread.join(timeout=5)
+        # YICHUAN: Close loop AFTER the join to ensure that the thread has already stopped the loop
+        self._loop.close()
+
+        if self._thread.is_alive():
+            raise Exception("Failed to stop runner thread")
+
+    def run_coro(self, coro: Coroutine) -> Any:
+        if not self._is_active():
+            raise Exception(f"Runner should be active before running coroutine: {coro}")
+
+        res = asyncio.run_coroutine_threadsafe(coro, loop=self._loop)
+        # YICHUAN: Waiting for the result will block, which is what we want for a sync wrapper
+        return res.result()
+
+
+# YICHUAN: CURRENTLY UNUSED, see EventLoopThreadRunner and explanation above
+def run_until_complete_safe(loop, coro) -> Any:
+    if loop.is_running():
+        # YICHUAN: Patching a loop multiple times is safe
+        nest_asyncio.apply(loop)
+    return loop.run_until_complete(coro)
+
+
+# YICHUAN: If we don't want to duplicate these raise  functions we can also modify them to directly take the error
+# status and reason instead of a response object (that way we can use the same for both sync and async)
+# For this commit I'm just adding separate functions for async so that I don't need to change too much code
+def raise_okta_unauthorized_error_async(
+    connection: SnowflakeConnection | None, response: aiohttp.ClientResponse
+) -> None:
+    Error.errorhandler_wrapper(
+        connection,
+        None,
+        DatabaseError,
+        {
+            "msg": f"Failed to get authentication by OKTA: {response.status}: {response.reason}",
+            "errno": ER_FAILED_TO_CONNECT_TO_DB,
+            "sqlstate": SQLSTATE_CONNECTION_REJECTED,
+        },
+    )
+
+
+def raise_failed_request_error_async(
+    connection: SnowflakeConnection | None,
+    url: str,
+    method: str,
+    response: aiohttp.ClientResponse,
+) -> None:
+    # YICHUAN: NO TELEMETRY FOR NOW
```

Comment:
I think these comments should be highlighted, I do this with `TODO` comments.
```suggestion
    # TODO: add async telemetry support
```